src/main/java/edu/harvard/iq/dataverse/api/Datasets.java

```diff
@@ -2502,6 +2503,65 @@ public Response addFileToDataset(@PathParam("id") String idSupplied,
     } // end: addFileToDataset
 
 
+    /**
+     * Clean storage of a Dataset
+     *
+     * @param idSupplied
+     * @return
+     */
+    @GET
+    @Path("{id}/cleanStorage")
+    public Response cleanStorage(@PathParam("id") String idSupplied) {
+        // get user and dataset
+        User authUser;
+        try {
+            authUser = findUserOrDie();
+        } catch (WrappedResponse ex) {
+            return error(Response.Status.FORBIDDEN,
+                    BundleUtil.getStringFromBundle("file.addreplace.error.auth")
+            );
+        }
+
+        Dataset dataset;
+        try {
+            dataset = findDatasetOrDie(idSupplied);
+        } catch (WrappedResponse wr) {
+            return wr.getResponse();
+        }
+        
+        // check permissions
+        if (!permissionSvc.permissionsFor(createDataverseRequest(authUser), dataset).contains(Permission.EditDataset)) {
+            return error(Response.Status.INTERNAL_SERVER_ERROR, "Access denied!");
+        }
+
+        List<String> deleted = new ArrayList<>();
+        Set<String> files = new HashSet<String>();
+        try {
+            for (DatasetVersion dv : dataset.getVersions()) {
+                for (FileMetadata f : dv.getFileMetadatas()) {
+                    String storageIdentifier = f.getDataFile().getStorageIdentifier();
+                    String location = storageIdentifier.substring(storageIdentifier.indexOf("://") + 3);
+                    String[] locationParts = location.split(":", 3);//separate bucket, swift container, etc. from fileName
+                    files.add(locationParts[locationParts.length-1]);
+                }
+            }
+            StorageIO<DvObject> datasetIO = DataAccess.getStorageIO(dataset);
+            List<String> allDatasetFiles = datasetIO.listAllFiles();
```

Comment:
One quick thought would be to just have a StoageIO.cleanDataset() method and keep the list and delete methods out of the public interface. There might be some circumstances where listing or deleting in the main code body might be useful, but I think it would be easier to maintain integrity if stores manage that directly. (FWIW: I think all stores currently use the same authority/identifier/filename format but they don't have to so putting the logic to identify valid files within the store would be important in that case.)

Reply:
I have done some refactoring. The thing that I am struggling right now is which files to keep. So I have a list of all files from all versions of the dataset (here called `files`) and right now I keep the files that start with any name in that list:
```
Predicate<String> filter = f -> {
  return files.stream().noneMatch(x -> f.startsWith(x));
};
```
I could use some help extending this filter to keep all files we want to keep. I can add an option to delete more files later on, if needed.